### PR TITLE
Line segment intersection

### PR DESCRIPTION
**Pull request by jbeenie**

### CODE REVIEW

File: PathKit/Sources/Line.swift [hunk 1]
```diff
@@ -9,21 +9,32 @@ import Foundation
...
+
     /// Determines whether `point` is on `self`.
     ///
     /// - Parameters:
     ///   - point: The point that may or may not be on `self`
     ///   - accuracy: The accuracy used to determine if the point lies on `self`.
     /// - Returns: True if `point` is on `self` within the given accuracy.
     public func contains(_ point: Point, accuracy: Double) -> Bool {
-        let dotProduct = Vector(head: point, tail: .zero) • Vector(angle: angle)
+        let dotProduct = Vector(head: point, tail: .zero) • Vector.unitVector(with: angle)
```
Comment (alejandro-isaza) [inline PathKit/Sources/Line.swift:37]: Better!

File: PathKit/Sources/LineSegment.swift [hunk 2]
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
...
 
-    /// Creates a line segment begins at the point `start` and ends the point `end`.
+    /// Creates a line segment beginning at `start` and ending at `end`.
     ///
     /// - Parameters:
     ///   - start: The starting point of the line segment.
     ///   - end: The end point of the line segment.
+    /// - Precondition: The `start` point and `end` point of the line segment must be different.
     public init(start: Point, end: Point) {
+        precondition(start != end)
```
Comment (alejandro-isaza) [inline PathKit/Sources/LineSegment.swift:24]: Why?

File: PathKit/Sources/LineSegment.swift [hunk 3]
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
...
+
+    public func offset(by vector: Vector) -> LineSegment {
+        return LineSegment(start: start + vector, end: end + vector)
+    }
+
+    /// Computes the intersection of two line segments.
+    /// Source: https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect
+    public static func intersection(_ lhs: LineSegment, _ rhs: LineSegment, accuracy: Double) -> Intersection {
+        let p = lhs.start
+        let q = rhs.start
```
Comment (alejandro-isaza) [inline PathKit/Sources/LineSegment.swift]: do we really need to add these `let`s?

File: PathKit/Sources/LineSegment.swift [hunk 4]
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
...
+        let qp = Vector(head: q, tail: p)
```
Comment (alejandro-isaza) [inline PathKit/Sources/LineSegment.swift]: Let's come up with a better name for this. Something like `headsVector`

File: PathKit/Sources/LineSegment.swift [hunk 5]
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
...
+        let r = Vector(head: lhs.end, tail: lhs.start)
+        let s = Vector(head: rhs.end, tail: rhs.start)
```
Comment (alejandro-isaza) [inline PathKit/Sources/LineSegment.swift]: `lhsVector` and `rhsVector`?

Comment (alejandro-isaza) [reply PathKit/Sources/LineSegment.swift]: Also why are they backwards?

File: PathKit/Sources/LineSegment.swift [hunk 6]
```diff
@@ -27,13 +27,64 @@ public struct LineSegment {
...
+            }
+        } else if parallel {
+            // lines are parallel and non-intersecting
+            return .empty
+        } else {
+            let rXs = r ✕ s
+            let t = qp ✕ s / rXs
+            let u = qp ✕ r / rXs
+
+            if zeroToOne.contains(t) && zeroToOne.contains(u) {
```
Comment (alejandro-isaza) [inline PathKit/Sources/LineSegment.swift]: I think the accuracy should be taken into account here?

File: PathKit/Sources/LineSegment.swift (hunk 5 above)
Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: They're not backwards, the vector points in the direction  tail -> head. So the tail is like the start of the vector and the head is like the end.

File: PathKit/Sources/LineSegment.swift (hunk 6 above)
Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: Good catch. Will fix

File: PathKit/Sources/LineSegment.swift (hunk 4 above)
Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: ok will do

File: PathKit/Sources/LineSegment.swift (hunk 3 above)
Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: will remove these

File: PathKit/Sources/LineSegment.swift (hunk 2 above)
Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: It avoids division by zero in a couple of places.

File: PathKit/Sources/LineSegment.swift (hunk 5 above)
Comment (alejandro-isaza) [reply PathKit/Sources/LineSegment.swift]: I get confused by this every time...

Comment (jbeenie) [reply PathKit/Sources/LineSegment.swift]: :)

**Comment by jbeenie** [comment]
I added the logic computing the overlap of intersecting LineSegments. Need to add units tests now to test that case and tests a couple other functions I added.

**Review by alejandro-isaza — APPROVED** [review]

**Pull request merged**